org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java

```diff
@@ -0,0 +1,197 @@
+import org.eclipse.xtext.xbase.typesystem.override.OverrideHelper;
+import org.eclipse.xtext.xbase.typesystem.override.ResolvedFeatures;
+import org.eclipse.xtext.xbase.typesystem.references.LightweightTypeReference;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.FluentIterable;
+import com.google.inject.Inject;
+
+/**
+ * @author Anton Kosyakov - Initial contribution and API
+ * @since 2.4
+ */
+public class XbaseCopyQualifiedNameService extends DefaultCopyQualifiedNameService {
+
+	@Inject
+	private OverrideHelper overrideHelper;
+
+	@Inject
+	private IBatchTypeResolver typeResolver;
+
+	protected String toQualifiedName(JvmExecutable jvmExecutable) {
+		return toFullyQualifiedName(jvmExecutable) //
+				+ "(" + toQualifiedNames(jvmExecutable.getParameters(), (parameter) -> parameter.getParameterType().getSimpleName()) + ")";
+	}
+
+	protected String toQualifiedName(XExpression expression, IResolvedExecutable resolvedExecutable, JvmExecutable executable,
+			IResolvedTypes resolvedTypes, List<XExpression> arguments) {
+		LightweightTypeReference actualType = resolvedTypes.getActualType(expression);
+		if (actualType != null && !actualType.isAny() && !actualType.isUnknown()) {
+			return actualType.getHumanReadableName();
+		}
+
+		int index = arguments.indexOf(expression);
+		if (resolvedExecutable == null) {
+			return executable.getParameters().get(index).getParameterType().getSimpleName();
+		}
+		return resolvedExecutable.getResolvedParameterTypes().get(index).getSimpleName();
+	}
+
+	@Override
+	public String getQualifiedName(EObject constructor, EObject constructorCall) {
+		if (constructor instanceof JvmConstructor && constructorCall instanceof XConstructorCall) {
+			return _getQualifiedName((JvmConstructor) constructor, (XConstructorCall) constructorCall);
+		} else if (constructor instanceof JvmExecutable && constructorCall instanceof XAbstractFeatureCall) {
+			return _getQualifiedName((JvmExecutable) constructor, (XAbstractFeatureCall) constructorCall);
+		} else if (constructor instanceof JvmExecutable && constructorCall != null) {
+			return _getQualifiedName((JvmExecutable) constructor, constructorCall);
+		} else if (constructor instanceof JvmExecutable && constructorCall == null) {
+			return _getQualifiedName((JvmExecutable) constructor, (Void) null);
+		} else if (constructor != null && constructorCall != null) {
+			return _getQualifiedName(constructor, constructorCall);
+		} else if (constructor != null && constructorCall == null) {
+			return _getQualifiedName(constructor, (Void) null);
+		} else if (constructor == null && constructorCall != null) {
+			return _getQualifiedName((Void) null, constructorCall);
+		} else if (constructor == null && constructorCall == null) {
+			return _getQualifiedName((Void) null, (Void) null);
+		} else {
+			throw new IllegalArgumentException("Unhandled parameter types: " + Arrays.asList(constructor, constructorCall).toString());
```

Comment:
You can remove this, it's impossible to end up here, I don't if you leave it though